src/hotspot/share/prims/agentList.cpp

```diff
@@ -0,0 +1,652 @@
+/*
+ * Copyright (c) 2023, Oracle and/or its affiliates. All rights reserved.
+ * DO NOT ALTER OR REMOVE COPYRIGHT NOTICES OR THIS FILE HEADER.
+ *
+ * This code is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License version 2 only, as
+ * published by the Free Software Foundation.
+ *
+ * This code is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
+ * version 2 for more details (a copy is included in the LICENSE file that
+ * accompanied this code).
+ *
+ * You should have received a copy of the GNU General Public License version
+ * 2 along with this work; if not, write to the Free Software Foundation,
+ * Inc., 51 Franklin St, Fifth Floor, Boston, MA 02110-1301 USA.
+ *
+ * Please contact Oracle, 500 Oracle Parkway, Redwood Shores, CA 94065 USA
+ * or visit www.oracle.com if you need additional information or have any
+ * questions.
+ */
+
+#include "precompiled.hpp"
+#include "prims/agentList.hpp"
+
+#include "cds/cds_globals.hpp"
+#include "jni.h"
+#include "jvmtifiles/jvmtiEnv.hpp"
+#include "prims/jvmtiEnvBase.hpp"
+#include "prims/jvmtiExport.hpp"
+#include "runtime/arguments.hpp"
+#include "runtime/atomic.hpp"
+#include "runtime/handles.inline.hpp"
+#include "runtime/interfaceSupport.inline.hpp"
+#include "runtime/java.hpp"
+#include "runtime/jniHandles.hpp"
+#include "runtime/os.inline.hpp"
+#include "runtime/thread.inline.hpp"
+#include "utilities/growableArray.hpp"
+
+Agent* AgentList::_list = nullptr;
+
+static inline Agent* head(Agent** list) {
+  assert(list != nullptr, "invariant");
+  return Atomic::load_acquire(list);
+}
+
+void AgentList::add(Agent* agent) {
+  assert(agent != nullptr, "invariant");
+  Agent* next;
+  do {
+    next = head(&_list);
+    agent->_next = next;
+  } while (Atomic::cmpxchg(&_list, next, agent) != next);
+}
+
+void AgentList::add(const char* name, char* options, bool absolute_path) {
+  add(new Agent(name, options, absolute_path));
+}
+
+void AgentList::add_xrun(const char* name, char* options, bool absolute_path) {
+  Agent* agent = new Agent(name, options, absolute_path);
+  agent->_is_xrun = true;
```

Comment:
Why direct access of private field instead of having a setter like other parts of the Agent API?

Reply:
I wanted all accesses to use the iterator. The only access is given to the iterator and AgentList by way of being friends. No need to expose more.